src/kubernetes_cluster/spec/controller/common.rs

```diff
@@ -17,7 +17,14 @@ pub struct ControllerState<K: ResourceView, R: Reconciler<K>> {
 
 pub struct OngoingReconcile<K: ResourceView, R: Reconciler<K>> {
     pub triggering_cr: K,
+    // Use two fields for pending k8s api request and pending external library result separately.
+    // Such design is not very reasonable because it doesn't make it apparent that the two fields
+    // are mutually exclusive, i.e., only one of them can be non-empty at some point.
+    // But this is fine, since we can describe the state as one of them is Option::Some(...) and
+    // the other is Option::None.
     pub pending_req_msg: Option<Message>,
+    pub pending_lib_req: Option<R::I>,
+    pub lib_response: Option<R::O>,
```

Comment:
Do we plan to put "controller issues a pending lib req" and "controller receives the lib response" in the same action? If so we probably don't need `pending_lib_req` as part of the state?

Reply:
Then how to make the exec code match the spec code?

Reply:
The spec `reconcile_core` still returns lib req (or k8s req), but you can "process" it and get the lib response in the same action (basically, make the interaction with external lib atomic). And you will need a spec version of "process".